### PR TITLE
Refactor selectors and add CenterNthSelector

### LINKED ISSUE #517: Mixin NthSelector
While working on `RadiusNthSelector` in #504 I duplicated a fair bit of code from `DirectionNthSelector`. I think it would be worth creating a `MixinNthSelector` class to hold this code and be a parent for `RadiusNthSelector`, `DirectionNthSelector`, and for a potential `LengthNthSelector`.

### LINKED ISSUE #530: Cannot select a hole in a chamfered object
For this example (slightly changed compared to https://github.com/CadQuery/cadquery/issues/522 to make clear circles do not have the same center)

```python
thickness = 2
height = 40
width = 65
diam = 4
tol = 0.05

l1, l2 = 50, 80
pts = [( 0,  0), ( 0, height/2), (l1, height/2 - 5), (l2, 0)]
upper_leg_hole = (l2 - 20, 0)

upper_leg = (cq.Workplane()
    .polyline(pts).mirrorX()
    .pushPoints([upper_leg_hole]).circle(diam/2 + tol).extrude(thickness)
    .edges("|Z and (not <X)").fillet(4)
)
upper_leg.faces(">Z").edges("%Circle")
```

I did not find a way to select the circle with CadQuery selectors

```python
upper_leg.faces(">Z").edges("%Circle").edges().objects

    [<cadquery.occ_impl.shapes.Edge at 0x7f24f0b94ee0>,
     <cadquery.occ_impl.shapes.Edge at 0x7f24f0b942b0>,
     <cadquery.occ_impl.shapes.Edge at 0x7f24f0cc73a0>,
     <cadquery.occ_impl.shapes.Edge at 0x7f24f0cc7ca0>]
```

image

`">X"` works

```python
upper_leg.faces(">Z").edges("%Circle").edges(">X").objects
  
   [<cadquery.occ_impl.shapes.Edge at 0x7f24f1be5130>]
```
image

`">X[0]"` fails

```python
upper_leg.faces(">Z").edges("%Circle").edges(">X[0]")

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-103-be1f63569cdd> in <module>
----> 1 upper_leg.faces(">Z").edges("%Circle").edges(">X[0]")

/opt/anaconda/envs/cq-ocp21/lib/python3.8/site-packages/cadquery/cq.py in edges(self, selector, tag)
    841         returns 4, because a cube has 4 edges parallel to the z direction
    842         """
--> 843         return self._selectObjects("Edges", selector, tag)
    844 
    845     def wires(

/opt/anaconda/envs/cq-ocp21/lib/python3.8/site-packages/cadquery/cq.py in _selectObjects(self, objType, selector, tag)
    737             else:
    738                 selectorObj = selector
--> 739             toReturn = selectorObj.filter(toReturn)
    740 
    741         return self.newObject(toReturn)

/opt/anaconda/envs/cq-ocp21/lib/python3.8/site-packages/cadquery/selectors.py in filter(self, objectList)
    726         Filter give object list through th already constructed complex selector object
    727         """
--> 728         return self.mySelector.filter(objectList)

/opt/anaconda/envs/cq-ocp21/lib/python3.8/site-packages/cadquery/selectors.py in filter(self, objectList)
    605             [+\|-\|<\|>\|] \<X\|Y\|Z>
    606         """
--> 607         return self.mySelector.filter(objectList)
    608 
    609 

/opt/anaconda/envs/cq-ocp21/lib/python3.8/site-packages/cadquery/selectors.py in filter(self, objectList)
    385 
    386         # choose the Nth unique rounded distance
--> 387         nth_distance = sorted(list(objectDict.keys()), reverse=not self.directionMax)[
    388             self.N
    389         ]

IndexError: list index out of range
```

### PR DESCRIPTION
**Pull request by marcus7070**
This PR will close #517 and add a CenterNthSelector to resolve #530. I touched a few other things as well, sorry for lumping them all in one PR but I've at least rebased all the commits into neat chunks.

---

Commit 6d2f3ba adds _NthSelector. It is an abstract class, it should only be inherited from, not instantiated.

Previously DirectionNthSelector lumped objects at a similar distance together by rounding that distance to a certain number of decimal places. This would mean that with a tolerance of 0.1, 0.249999999 and 0.250000001 would be put into seperate bins. So I've changed it to a simple clustering method. It's not perfect, but it's simple and I think it's an improvement over the previous one.

These two commits should not change or add any behaviour to CQ (except to remove that rounding issue). All our previous tests pass with these two new commits.

---

344651d added CenterNthSelector, to select the Nth object/objects along a direction. Similar to DirectionNthSelector, but without running them through ParallelDirSelector first.

Now DirectionMinMaxSelector can become a simple subclass of CenterNthSelector with the index set to -1.

---

Previously "Combining Selectors" appeared as a top level item on the side bar menu.
screenshot2020-12-18-111717

After a242473:
screenshot2020-12-18-154652

---

While looking at the docs I noticed the tables were overflowing and it was hard to read. 

screenshot2020-12-18-154812

4486381 adds some extra CSS from [here](https://github.com/readthedocs/sphinx_rtd_theme/issues/117) to change this behaviour, now tables (such as the Class Summary) line wrap. I've quickly checked all the pages in the HTML docs and I can't see any problems, I think it's only improved things. Now:

screenshot2020-12-18-154905

---

I've tried to add mypy type hints throughout all this, but I'm going to have to ask for help from the mypy gurus for one bit. I can't see why mypy errors on `RadiusNthSelector.filter`. I'm leaving this as a draft PR until I get the mypy errors sorted.

### CODE REVIEW

**Comment by marcus7070** [comment]
Hey GitHub, I said this will close #531 in my commit message, pay attention!

File: cadquery/selectors.py [hunk 1]
```diff
@@ -155,287 +154,307 @@ def isInsideBox(p):
...
-                normal = o.normalAt(None)
-
-                if self.test(normal):
-                    r.append(o)
-            elif type(o) == Edge and (
+            if isinstance(o, Face):
+                # a face is only parallel to a direction if it is a plane, and its normal is parallel to the dir
+                test_vector = o.normalAt(None)
+            elif isinstance(o, Edge) and (
                 o.geomType() == "LINE" or o.geomType() == "PLANE"
```
Comment (marcus7070) [inline cadquery/selectors.py:181]: @adam-urbanczyk, I see you wrote this line 4 years ago. I don't think an Edge can have `geomType() == "PLANE"`. Was this a typo or am I wrong?

**Comment by marcus7070** [comment]
Thanks for the mypy stuff @adam-urbanczyk.

codecov is right, there is some checks I've missed, I'm working on some more tests now. Once I've got them done I think this is ready for review and I'll change from draft to regular pull request.

File: cadquery/selectors.py (hunk 1 above)
Comment (adam-urbanczyk) [reply cadquery/selectors.py:181]: Looks indeed like a mistake.

Comment (marcus7070) [reply cadquery/selectors.py:181]: OK, I'll fix it and make a test for it. I'll write a more detailed comment when I'm done.

**Comment by marcus7070** [comment]
Commit e19a05d is breaking under certain circumstances.

I think it was always the intent for ParallelDirSelector to filter out any non-planar surface, but there was a typo where the check was performed on Edges and any Face was let through as long as it's `.tangentAt(None)` was correct. For instance:
```python
import cadquery as cq

part = (
    cq
    .Workplane()
    .box(10, 10, 10)
    .faces(">X")
    .workplane()
    .hole(1)
)
```
Before selection:
screenshot2020-12-08-162315

`.faces(">X[1]")` and `">Z[1]"` select the side of the cube, as you expect. `.faces(">Y[1]")` catches the cylinder, because it's normal happens to be in right direction.

screenshot2020-12-08-162330

So after e19a05d, if someone had made a model like above, used DirectionMinMaxSelector to select a planar face and caught a non-planar edge with the correct normal, their ">Y[2]" would now have to become "Y[1]". 

---

This is ready for review now.

**Comment by adam-urbanczyk** [comment]
Note to self: check if this is compatible with downstream applications, e.g.

https://github.com/michaelgale/cq-kit/blob/master/cqkit/cq_selectors.py

**Comment by marcus7070** [comment]
> Note to self: check if this is compatible with downstream applications, e.g.
> 
> https://github.com/michaelgale/cq-kit/blob/master/cqkit/cq_selectors.py

Great point, I did not check this.

**Comment by adam-urbanczyk** [comment]
@marcus7070 do you want to add the `CenterNthSelector` to the string syntax? Maybe as `>>`?

**Comment by marcus7070** [comment]
> Note to self: check if this is compatible with downstream applications, e.g.
> 
> https://github.com/michaelgale/cq-kit/blob/master/cqkit/cq_selectors.py

I just got the cq-kit tests to run with this branch. I get failures in cq-kit's file exporters related to a missing `defaultdict`, but all the selector tests pass.

```
collected 36 items                                                                                                                     

test_discrete.py ...                                                                                                             [  8%]
test_files.py .FFFF..                                                                                                            [ 27%]
test_geometry.py ...............                                                                                                 [ 69%]
test_pprint.py .                                                                                                                 [ 72%]
test_selectors.py ..........                                                                                                     [100%]

========================== short test summary info ==========================
FAILED test_files.py::test_step_export_simple - NameError: name 'defaultdict' is not defined
FAILED test_files.py::test_step_export - NameError: name 'defaultdict' is not defined
FAILED test_files.py::test_export_function - NameError: name 'defaultdict' is not defined
FAILED test_files.py::test_step_import - NameError: name 'defaultdict' is not defined
========================= 4 failed, 32 passed in 0.93s =========================
```

**Comment by marcus7070** [comment]
> @marcus7070 do you want to add the `CenterNthSelector` to the string syntax? Maybe like so: `>>X[0]`?

Will do, I'll try to get to this tomorrow.

**Comment by adam-urbanczyk** [comment]
Thanks for checking @marcus7070 . @michaelgale do you have anything against this PR?

**Comment by adam-urbanczyk** [comment]
OK I added the implementation and tests for `>>` and `<<`. NB: `>>X[-1]` is the same as `>>X` and `>X`.

**Comment by michaelgale** [comment]
@adam-urbanczyk This PR looks fine to me!  Since I do not make use of the indexed Nth selectors, the cq-kit selectors are not affected.  I've added an import for defaultdict to clear the errors noted above.

**Comment by michaelgale** [comment]
@adam-urbanczyk Incidentally, I noticed cq-kit now fails many CI tests.  It looks like the new CQ2.x that gets pulled into the test environment breaks a lot of the my direct access into OCP.  e.g.:

```
>           writer.Transfer(self.shape.val().wrapped, STEPControl_AsIs)
1050E           TypeError: in method 'STEPControl_Writer_Transfer', argument 2 of type 'TopoDS_Shape const &'
1051
1052../../../../miniconda/envs/cadquery/lib/python3.7/site-packages/cqkit-0.2.0-py3.7.egg/cqkit/cq_files.py:324: TypeError
```

They are almost all type errors--is it the `const` reference?

**Comment by adam-urbanczyk** [comment]
I cannot reproduce it locally @michaelgale , maybe you can open a new issue with more detailed info?

**Review by adam-urbanczyk — APPROVED** [review]

**Comment by adam-urbanczyk** [comment]
I added a #559 fix to this PR

**Comment by michaelgale** [comment]
@adam-urbanczyk The CI issues with CQ-Kit are resolved--no more errors.  My travis environment was outdated and was pulling in pythonocc-core and oce, rather than the latest OCP based OCCT + cadquery.

**Comment by adam-urbanczyk** [comment]
@jmwright are you going to take a look? I think it can be merged now.

**Comment by jmwright** [comment]
I'll try to take a look here in a couple of hours and then merge.

**Comment by marcus7070** [comment]
> OK I added the implementation and tests for `>>` and `<<`. NB: `>>X[-1]` is the same as `>>X` and `>X`.

Thanks for taking care of that @adam-urbanczyk. I have never used pyparsing before and I wasn't looking forward to learning it.

I've just got one line in the docs that needs fixing.

File: cadquery/selectors.py [hunk 2]
```diff
@@ -38,24 +37,24 @@
...
-            :param objectList: list to filter
-            :type objectList: list of FreeCAD primatives
-            :return: filtered list
-
-            The default implementation returns the original list unfiltered
+        Filter the provided list.
 
+        The default implementation returns the original list unfiltered.
+        :param objectList: list to filter
+        :type objectList: list of FreeCAD primatives
```
Comment (jmwright) [inline cadquery/selectors.py:56]: Might as well change "FreeCAD" from the original here to OCCT. Also, "primatives" from the original should be "primitives".

File: doc/selectors.rst [hunk 3]
```diff
@@ -7,53 +7,63 @@ String Selectors Reference
 CadQuery selector strings allow filtering various types of object lists. Most commonly, Edges, Faces, and Vertices are
 used, but all objects types can be filtered.
 
-String selectors are simply shortcuts for using the full object equivalents. If you pass one of the
-string patterns in, CadQuery will automatically use the associated selector object.
+Object lists are created by using the following methods, which each each collect a type of shape:
```
Comment (jmwright) [inline doc/selectors.rst:10]: Two instances of "each" in this line.

**Review by jmwright — APPROVED** [review]
Just a couple of minor comments. Thanks for all the work you've put into this @marcus7070 !

**Comment by jmwright** [comment]
Doc strings fixed, merging. Thanks @marcus7070 !

**Pull request merged**